### PR TITLE
Sam Landsman

### PR DESCRIPTION
**Pull request by steve2020**

### CODE REVIEW

File: Answers.md [hunk 1]
```diff
@@ -1,6 +1,25 @@
 # Your responses to the short answer questions should be laid out here using Mark Down.
 1. Describe the biggest difference between `.forEach` & `.map`.
+
+forEach` loops over each element in an array, but does not automatically create a new array; while map does create a new a array corresponsing to
+each eelemnt
+
```
Comment (JasonCorp84) [inline Answers.md]: 👍

File: Answers.md [hunk 2]
```diff
@@ -1,6 +1,25 @@
...
 2. What is the difference between a function and a method?
+
+a method refers to a function within an object (an object's functions).
+
```
Comment (JasonCorp84) [inline Answers.md]: 👍

File: Answers.md [hunk 3]
```diff
@@ -1,6 +1,25 @@
...
 3. What is closure?
-4. Describe the four rules of the 'this' keyword.
-5. Why do we need super() in an extended class?
\ No newline at end of file
+
+a combination of a function and the lexical environment within which that function was declared.
+
```
Comment (JasonCorp84) [inline Answers.md]: 👍

File: Answers.md [hunk 4]
```diff
@@ -1,6 +1,25 @@
...
+1. when 'this' is in global scope, it refrences the window or console object.
+2. implicit binding: when object.function is called, 'this' (in the function) refers to the the object preceding the dot.
+3. new binding: when the 'new' keyword is used to create an instance of an object, 'this' in the constructor funtion refers to that 
+    object instance.
+4. explicit binding: the 'call' and 'apply' methods explicitly state which object 'this' refers to.
+
+5. Why do we need super() in an extended class?
+
+to cause a child class to inherit the properties and methods of its parent class.
```
Comment (JasonCorp84) [inline Answers.md]: 👍 👍 👍 👍

**Comment by mixelpixel** [comment]
@steve2020 👋

It's nice to see that you are committing frequently. You get a gold ⭐️ !! Glad to see you are getting comfortable with the Git and GitHub tools :octocat: 🎉 

Looking ahead to the larger projects you'll be working on, I'd ❤️ to see your msgs conveying more information, more frequently. This habit will make you a better programmer! Any time you complete an assignment, a function, add a component, or try a new feature in your programs, if you make note of the updates by using Git commit messages, your Git history will _**tell the story of your software engineering adventure!**_

While the Git commit titles are short, you can add a "body" to the messages which can be much more detailed. BEFORE you add the closing quote to the commit message, hit enter twice to add a new line between the title and the body. For example:

```
$ git commit -m "This is a Git commit TITLE less than 72 characters

This is a Git commit BODY which can contain as much pertinent information and descriptions as you see fit. Some development teams like tons and tons of information, and depending on the scope of the commit, it may be necessary to make lots of detailed notes - especially in projects like React where there can be a vast amount of files and a complex project architecture.
- I did this.
- I did that.
- this happens.
- that needs work.
- this thing got refactored.
- where's my Lambda swag?????"
```

...on GitHub in the Pull Requests, the body will be accessible by clicking on the three little dots:

triple-dots

...such that they will reveal the BODY:

reveal

If you're not already in the #git channel in Slack, come check it out: https://lambdaschoolstudents.slack.com/archives/C5Z8BMX6K/p1523565197000063

😺

File: challenges/objects-arrays.js [hunk 5]
```diff
@@ -6,12 +6,55 @@
...
+
+class Tyrannosaurus extends Dinosaur {
+  constructor(attributes) {
+    super(attributes);
+  }
+  roar() {
+      return `RAWERSRARARWERSARARARRRR!`;
+  }
+}
+
```
Comment (JasonCorp84) [inline challenges/objects-arrays.js]: 👍

File: challenges/objects-arrays.js [hunk 6]
```diff
@@ -6,12 +6,55 @@
...
 // velociraptor, carnivorous, 15kg, 1.8m, Late Cretaceious
 
+const velociraptor = new Dinosaur({
+  name: 'velociraptor',
+  diet: 'carnivorous',
+  weight: '15kg',
+  length: '1.8m',
+  period: 'Late Cretaceious'
+});
+
```
Comment (JasonCorp84) [inline challenges/objects-arrays.js]: 👍 👍 👍

File: challenges/objects-arrays.js [hunk 7]
```diff
@@ -46,6 +90,11 @@ const graduates = [{"id":1,"first_name":"Cynde","university":"Missouri Southern
 
 Once you have the new array created, sort the universities alphabetically and log the result. */
 const universities = [];
+
+for (i = 0; i < graduates.length; i++) {
+  universities.push(graduates[i].university);
+}
+
 console.log(universities)
```
Comment (JasonCorp84) [inline challenges/objects-arrays.js]: Lets not forgot to sort by alphabet

File: challenges/objects-arrays.js [hunk 8]
```diff
@@ -55,11 +104,27 @@ Name email@example.com
...
+
+for (i = 0; i < graduates.length; i++) {
+  contactInfo.push(graduates[i].first_name);
+}
+
+for (i = 0; i < contactInfo.length; i++) {
+  contactInfo[i] += ' ' + graduates[i].email;
+}
+
 console.log(contactInfo);
```
Comment (JasonCorp84) [inline challenges/objects-arrays.js]: 👍

File: challenges/objects-arrays.js [hunk 9]
```diff
@@ -55,11 +104,27 @@ Name email@example.com
...
 
 /* Request 3: Find out how many universities have the string "Uni" included in their name. Create a new array called uni that contains them all. Log the result. */
 const uni = [];
+
+for (i = 0; i < graduates.length; i++) {
+  if (graduates[i].university.includes("Uni")) {
+    uni.push(graduates[i].university);
+  }
+}
+
```
Comment (JasonCorp84) [inline challenges/objects-arrays.js]: ok, but how many? :)

File: challenges/objects-arrays.js [hunk 10]
```diff
@@ -80,8 +145,22 @@ zooAnimals = [{"animal_name":"Jackal, asiatic","population":5,"scientific_name":
 
 // The zoos need a list of all their animal's names converted to lower case.  Create a new array named lowerCase and map over each name to convert them all to lower case.  Log the resut.
 let lowerCase = [];
+
+lowerCase = zooAnimals.map(animal => animal.animal_name.toLowerCase());
+
 console.log(lowerCase);
```
Comment (JasonCorp84) [inline challenges/objects-arrays.js]: 👍

File: challenges/objects-arrays.js [hunk 11]
```diff
@@ -80,8 +145,22 @@ zooAnimals = [{"animal_name":"Jackal, asiatic","population":5,"scientific_name":
...
+zooAnimals.forEach(function(animal) {
+  populationTotal.push(animal.population);
+});
+console.log(populationTotal);
+
+let populationSum = populationTotal.reduce(function (accumulator, currentValue) {
+  return accumulator + currentValue;
+}, 0);
+
+console.log(populationSum);
```
Comment (JasonCorp84) [inline challenges/objects-arrays.js]: 👍

File: challenges/functions.js [hunk 12]
```diff
@@ -7,26 +7,42 @@
...
+}
+
+function multiply(num1, num2) {
+  return num1 * num2;
+}
+ 
+function greeting(fName, lName) {
+  return `Hello ${fName} ${lName}, nice to meet you!`;
+}
+
```
Comment (JasonCorp84) [inline challenges/functions.js]: 👍 👍 👍

File: challenges/functions.js [hunk 13]
```diff
@@ -7,26 +7,42 @@
...
 
 
 // ==== Closures ==== 
 // Question 1: Explain in your own words why the example below is a closure.
 
-// Explanation: 
+// Explanation: because we are calling a function on a variable outside the function in which the the variablw was declared.
 
 // Question 2: Given the example below, what scope is the external variable in?
+// global scope
```
Comment (JasonCorp84) [inline challenges/functions.js]: 👍 👍

File: challenges/prototypes.js [hunk 14]
```diff
@@ -6,28 +6,46 @@
   Create a constructor function named CuboidMaker that accepts properties for length, width, and height
 */
 
+function CuboidMaker(props) {
+  this.length = props.length;
```
Comment (JasonCorp84) [inline challenges/prototypes.js]: let's not forgot the constructor

File: challenges/prototypes.js [hunk 15]
```diff
@@ -6,28 +6,46 @@
...
 /* == Step 2: Volume Method ==
   Create a method using CuboidMaker's prototype that returns the volume of a given cuboid's length, width, and height
   
   Formula for cuboid volume: length * width * height
 */
 
+CuboidMaker.prototype.volume = function() {
+  return this.length * this.width * this.height;
+};
+
```
Comment (JasonCorp84) [inline challenges/prototypes.js]: 👍

File: challenges/prototypes.js [hunk 16]
```diff
@@ -6,28 +6,46 @@
...
 
 /* == Step 3: Surface Area Method ==
   Create another method using CuboidMaker's prototype that returns the surface area of a given cuboid's length, width, and height. 
 
   Formula for cuboid surface area of a cube: 2 * (length * width + length * height + width * height)
 */
 
+CuboidMaker.prototype.surfaceArea = function() {
+  return 2 * (this.length * this.width + this.length * this.height + this.width * this.height);
+};
```
Comment (JasonCorp84) [inline challenges/prototypes.js]: 👍

File: challenges/prototypes.js [hunk 17]
```diff
@@ -6,28 +6,46 @@
...
   Create a cuboid object that uses the new keyword to use our CuboidMaker constructor
   Add properties and values of length: 4, width: 5, and height: 5 to cuboid.   
 */
 
+const cuboid = new CuboidMaker({
+  length: 4,
+  width: 5,
+  height: 5
+});
+
```
Comment (JasonCorp84) [inline challenges/prototypes.js]: 👍

File: challenges/classes.js [hunk 18]
```diff
@@ -1,11 +1,64 @@
...
+// CuboidMaker.prototype.surfaceArea = function() {
+//   return 2 * (this.length * this.width + this.length * this.height + this.width * this.height);
+// };
+
+class CuboidMaker {
+  constructor(props) {
+    this.length = props.length;
+    this.width = props.width;
+    this.height = props.height;
+  }
```
Comment (JasonCorp84) [inline challenges/classes.js]: 👍

File: challenges/classes.js [hunk 19]
```diff
@@ -1,11 +1,64 @@
...
+  volume() {
+    return this.length * this.width * this.height;
+  }
+  surfaceArea() {
```
Comment (JasonCorp84) [inline challenges/classes.js]: 👍

File: challenges/classes.js [hunk 20]
```diff
@@ -1,11 +1,64 @@
...
+    return 2 * (this.length * this.width + this.length * this.height + this.width * this.height);
+  }
+}
+
+const cuboid = new CuboidMaker({
+  length: 4,
+  width: 5,
+  height: 5
+});
```
Comment (JasonCorp84) [inline challenges/classes.js]: 👍

File: challenges/classes.js [hunk 21]
```diff
@@ -1,11 +1,64 @@
...
+  }
+  //volume formula is the same as cuboid
+}
+
+const cube = new CubeMaker({
+  length: 4,
+  width: 4,
+  height: 4
+});
+
```
Comment (JasonCorp84) [inline challenges/classes.js]: 👍

**Comment by JasonCorp84** [comment]
Sam,

### Self Study Questions
[x]  Answered all 

### Challenges
[ x] Objects 
  [x ] Create 3 dinosaur objects
  [ x] Test your objects using console logs
  [ x] Create rawr method

[ x] Arrays
  [ x] Part 1 - Complete all requests without using methods
  [ x] Part 2 - Complete all requests using methods

[ x] Functions
  [ x] Callbacks
  [ x] Closure

[ x] Prototypes
  [x ] Base Constructor
  [ x] Volume Method
  [ x] Surface Area Method
  [ x] Create new object that uses CuboidMaker

[ x] Classes
  [ x] Bring in prototype
  [ x] Test using console logs
  [ x] Extend base class with sub class

## Good Habits
  [ x] Pull Request title is your first and last name (add any notes in the comment field)
  [ x] Code has been committed regularly
  [ x] Commit comments are informative
  [ x] Code is organized in a logical way

Great job on your assignment. I like your clean concise explanations and code. 
Keep up the good work!
{3}

**Pull request closed without merging**